### ISSUE TITLE
Use resources more intelligent

### ISSUE THREAD

**Issue by davidak** [NixOS/nix#4137]
**Is your feature request related to a problem? Please describe.**
Build fails because `/tmp` has not enough space (only 24GB in this case) https://github.com/NixOS/nixpkgs/issues/54707

**Describe the solution you'd like**
Nix builds multiple packages at once. Maybe the build would not fail when Nix only builds Chromium and not also twemoji-color-font.

**Comment by davidak** [comment]
## Workaround

Instead of

`nixos-rebuild -I nixpkgs=https://github.com/worldofpeace/nixpkgs/archive/pantheon-stuff-20.09.tar.gz boot`

build just one package

`nix-build 'https://github.com/worldofpeace/nixpkgs/archive/pantheon-stuff-20.09.tar.gz' -A chromium`

but that also fails https://github.com/NixOS/nixpkgs/issues/100281

**Comment by edolstra** [comment]
Use `-j1`? It would be hard for Nix to predict build-time resource requirements of packages.

**Comment by zimbatm** [comment]
Nix should always use the disk for its tmpdir. Eg: `/nix/store/tmp/<build-id>`, mapped to /tmp inside of the build.

Using the host tmpdir has too much variance and is causing a lot of build and reproducibility issues.

**Comment by davidak** [comment]
In general, building in ramdisk is nice for performance reasons and the disk is not used heavily. It would be nice if it is possible to keep that, in most cases.

**Comment by flokli** [comment]
I agree with @davidak, we shouldn't enforce people to run builds on disks, if they have enough RAM available.

A more "canonical" location for build artifacts would be `/var/tmp`, like on gentoo, which is less likely to be RAM-backed (but people can opt in to do so, by creating the tmpfs mount by themselves).

**Comment by zimbatm** [comment]
Is there a way to turn off that optimization? I think it's problematic if builds are not repeatable across systems, it makes Nix less reliable.

**Comment by flokli** [comment]
> Is there a way to turn off that optimization? I think it's problematic if builds are not repeatable across systems, it makes Nix less reliable.

Not sure I understand. Putting `/tmp` on tmpfs is quite common, and true on NixOS when `boot.tmpOnTmpfs` is set to true (defaults to false).

Gentoo uses `/var/tmp` for builds, because it's more likely to be backed on disk: https://unix.stackexchange.com/a/30504. Nix should probably do so, too.

I don't think Nix should try to outsmart the way people set up mount points. If users decide to put the folder Nix temporarily builds inside on a tmpfs (if they happen to have a lot of RAM), or some networked storage, that's their decision :-)

Back to the original post: There's not much Nix can do if the location Nix is building in is running out of disk space (or inodes).

Reducing the amount of concurrent builds might be a way to migitate, or making that location configurable. Defaulting it to `/var/tmp/…` should make all this less likely to appear.

**Comment by davidak** [comment]
>It would be hard for Nix to predict build-time resource requirements of packages.

Yes, but we know which packages cause problems. Can we add a metadata attribute for needed storage and let Nix fail when that storage is not available? Also have a parameter to ignore the warning (for testing etc).

From my head:

- chromium
- llvm
- libreoffice

linux and firefox is actually fine with 15 GB (32 GB RAM)

we could document that we assume the user to have 15 GB free build space and if a package needs more, we add a warning and a hint to assign more ram to /tmp or put it on disk

When a package is known to need 15 GB and the build location has only 16 GB available, Nix should only start this one build and not more

>There's not much Nix can do if the location Nix is building in is running out of disk space (or inodes).

Nix can know before if it will happen and not waste some hours of computation time and fail directly. Don't cause unnecessary issues for users!

Right now Nix is dumb and runs happily into every issue. It could detect before if it has the needed resources to build a package successfully. That would make it a lot more user friendly!

**Comment by hw-lunemann** [comment]
The way nixos sets up and uses /tmp also breaks with https://www.freedesktop.org/software/systemd/man/file-hierarchy.html
as well as https://refspecs.linuxfoundation.org/FHS_3.0/fhs-3.0.html#purpose50 and https://refspecs.linuxfoundation.org/FHS_3.0/fhs-3.0.html#purpose17

I personally have run into problems before because I was expecting nixos to follow these recommendations.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
